### PR TITLE
feat(gpt-gateway): add logic code to connect openai service

### PR DESCRIPTION
**Pull request by morhag90**
Close: #6796 

## PR Details

Add the custom code for kafka connection and connect to openai service.

### Summary
🔄🗣️🧠

This pull request adds the functionality to start a GPT chat conversation based on a kafka message input, using the conversation type, template, and openai services. It also updates the environment variables and the tsconfig.app.json file to support the new features. The main files affected are `kafka.controller.ts`, `conversationType.service.ts`, and `template.service.ts`.

> _`Template` and `kafka`_
> _Chat with openai service_
> _Spring of conversations_

### Walkthrough
*  Change the port and server URL of the admin app to avoid conflicts with the gateway app ([link](https://github.com/amplication/amplication/pull/6867/files?diff=unified&w=0#diff-aa5c146201dd31a2794dd1ceb30c52cb01e9d8c446c840bc35c8e7a2d9cb5c44L1-R1))
*  Copy the environment variables of the gateway app to the admin app, and use a different database name for the admin app ([link](https://github.com/amplication/amplication/pull/6867/files?diff=unified&w=0#diff-efb9bf1f9592f875db23f69fa7de17cfe43bd8902a8ae4874fe9c040abfd2cadL3-R7))
*  Import the `template` module to the `conversationType` module, to enable the use of the `template` service in the `conversationType` service ([link](https://github.com/amplication/amplication/pull/6867/files?diff=unified&w=0#diff-476178891b9f9b285b12b573beedb097f24a30f54c4d7b133b43fda6aaf983ecL6-R13))
*  Modify the `conversationType` service to inject the `kafkaProducer` service and the `template` service as dependencies, and add a new method `startConversation` to start a conversation based on a message input ([link](https://github.com/amplication/amplication/pull/6867/files?diff=unified&w=0#diff-eda5b0a02117eff188d4ead265117f9520e853bff47022443663edc9dcfbfdacL2-R40))
*  Define new input types `MessageInput` and `MessageParam` to represent the message input that triggers a conversation, and the parameters for the template ([link](https://github.com/amplication/amplication/pull/6867/files?diff=unified&w=0#diff-cf65a37b86e064f753d72814e1d2e014e7f7951afd3a6b2cd43a9bbd15f76a53R1-R20), [link](https://github.com/amplication/amplication/pull/6867/files?diff=unified&w=0#diff-5d8a792766113994265158a63265c3d005cc37f42a762c4f946e40d826de099aR1-R16))
*  Inject the `conversationType` service to the `kafka` controller as a dependency, and modify the event handler for the `triggergptchatcompletion` topic to parse the message input from the kafka message value, and call the `startConversation` method with the message input ([link](https://github.com/amplication/amplication/pull/6867/files?diff=unified&w=0#diff-79d1c1d15638cda422a5282abc8f7049a75edebab15706dec24f62c8852ee6d1L4-R10), [link](https://github.com/amplication/amplication/pull/6867/files?diff=unified&w=0#diff-79d1c1d15638cda422a5282abc8f7049a75edebab15706dec24f62c8852ee6d1L12-R17))
*  Import the `conversationType` module to the `kafka` module, to enable the use of the `conversationType` service in the `kafka` controller ([link](https://github.com/amplication/amplication/pull/6867/files?diff=unified&w=0#diff-7eeda3bc299d3efee2ee54cad47994075b7564758c6eaed5d26c1509113aa24bL7-R11))
*  Define new args type `ProcessTemplateArgs` and input type `ProcessTemplateInput` to represent the arguments and input for the `processTemplate` mutation, containing the template ID and the parameters for the template ([link](https://github.com/amplication/amplication/pull/6867/files?diff=unified&w=0#diff-e2bc0fc094315fb24c2230e898f93e60c56899e1b8c40691698b10afe8d0030bR1-R10), [link](https://github.com/amplication/amplication/pull/6867/files?diff=unified&w=0#diff-4f966ec180591f2bbf9bcad2fdf0bd0b9fefa62edff85c74252c2e4386fd77a6R1-R15))
*  Import the `openai` service to the `template` module, to enable the use of the `openai` service in the `template` service ([link](https://github.com/amplication/amplication/pull/6867/files?diff=unified&w=0#diff-bba7982ffa15041e1a0ffd51f87cd7f0aef4e64163aa233a778d54cb20d0449fL6-R10))
*  Modify the `template` service to inject the `openai` service as a dependency, and add a new method `processTemplate` to process a template message based on the input, using the `openai` service to create a chat completion ([link](https://github.com/amplication/amplication/pull/6867/files?diff=unified&w=0#diff-8aa2ceeac929944839665f1d5593d652bb65f4f3bd039fedfd9f2e83c715e99cL4-R70))
*  Add the `experimentalDecorators` option to the `tsconfig.app.json` file, to enable the use of the `@EventPattern` decorator in the `kafka` controller ([link](https://github.com/amplication/amplication/pull/6867/files?diff=unified&w=0#diff-b47d53d1d638bce023ebef0f0eab3638ad14815a14ab29ed0f4ad01fa1eeb2ebL8-R9))

## PR Checklist

- [x] Tests for the changes have been added
- [x] `npm test` doesn't throw any error

### CODE REVIEW

File: packages/amplication-gpt-gateway/src/conversationType/conversationType.service.ts [hunk 1]
```diff
@@ -1,10 +1,41 @@
...
+    protected readonly templateService: TemplateService
+  ) {
     super(prisma);
   }
+
+  async startConversion(message: MessageInput): Promise<void> {
+    const { messageTypeId, params, requestUniqueId } = message;
+    const messageType = await this.prisma.conversationType.findUnique({
+      where: {
+        id: messageTypeId,
```
Comment (yuval-hazaz) [inline packages/amplication-gpt-gateway/src/conversationType/conversationType.service.ts:23]: we need to look for the conversation type by key and not by ID

```suggestion
        key: messageTypeKey,
```

File: packages/amplication-gpt-gateway/src/conversationType/conversationType.service.ts [hunk 2]
```diff
@@ -1,10 +1,41 @@
...
+    });
+
+    if (!messageType || !messageType.templateId)
+      throw new Error("Template not found");
+
+    const result = await this.templateService.processTemplateMessage({
+      templateId: messageType.templateId,
+      params,
+    });
+    if (!result) throw new Error("Template not found");
```
Comment (yuval-hazaz) [inline packages/amplication-gpt-gateway/src/conversationType/conversationType.service.ts:34]: I dont think we need to throw... instead we need to return the error to Amplication Server so the user will know that the process failed

File: packages/amplication-gpt-gateway/src/dto/MessageInput.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
+import { Field, InputType } from "@nestjs/graphql";
+import { MessageParam } from "../dto/MessageParam";
+
+@InputType()
+class MessageInput {
```
Comment (yuval-hazaz) [inline packages/amplication-gpt-gateway/src/dto/MessageInput.ts:5]:
```suggestion
class StartConversationInput {
```

File: packages/amplication-gpt-gateway/src/dto/MessageInput.ts [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+  @Field(() => String, {
+    nullable: true,
+  })
+  requestUniqueId!: string;
+
+  @Field(() => String, {
+    nullable: true,
+  })
+  messageTypeId!: string;
```
Comment (yuval-hazaz) [inline packages/amplication-gpt-gateway/src/dto/MessageInput.ts:14]:
```suggestion
  messageTypeKey!: string;
```

File: packages/amplication-gpt-gateway/src/kafka/kafka.controller.ts [hunk 5]
```diff
@@ -1,14 +1,19 @@
...
 import { Controller } from "@nestjs/common";
+import { MessageInput } from "../dto/MessageInput";
+import { ConversationTypeService } from "../conversationType/conversationType.service";
 
 @Controller("kafka-controller")
 export class KafkaController {
+  constructor(protected readonly conversationType: ConversationTypeService) {}
+
   @EventPattern("triggergptchatcompletion")
   async onTriggergptchatcompletion(
```
Comment (yuval-hazaz) [inline packages/amplication-gpt-gateway/src/kafka/kafka.controller.ts:12]: Update the Topic name on amplication to GptConversationStart (and  GptConversationComplete for the other side)

**Review by yuval-hazaz — CHANGES_REQUESTED** [review]

File: packages/amplication-gpt-gateway/src/conversationType/conversationType.service.ts (hunk 2 above)
Comment (morhag90) [reply packages/amplication-gpt-gateway/src/conversationType/conversationType.service.ts:34]: @yuval-hazaz  I looked at other implementations in our repo, and they all catch the error and write to the log. Does that sound like a good solution, or do you think that we should add it to the response ( ass status message)?

**Review by yuval-hazaz — CHANGES_REQUESTED** [review]

**Review by yuval-hazaz — APPROVED** [review]

**Pull request merged**